service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/support/ShareCaseController.java

```diff
@@ -45,9 +56,52 @@ public void shareCase(@PathVariable("caseId") String caseId, @RequestBody Users
                 new CaseUser(userId, CASE_ROLES)));
     }
 
+    @PostMapping(value = "/support/cases/share", consumes = APPLICATION_JSON_VALUE)
+    @Secured("caseworker-publiclaw-systemupdate")
+    public void shareCases(@Valid @RequestBody ShareLocalAuthorityCases shareCases) {
+
+        final String query = new JSONObject()
+            .put("from", 0)
+            .put("size", 1000)
+            .put("query", of("match", of("data.caseLocalAuthority", shareCases.localAuthority)))
+            .toString();
+
+        final List<CaseDetails> casesDetails = coreCaseDataService.searchCases(CASE_TYPE, query).stream()
+            //`Extra precaution as ES match does not guarantee strict equality
+            .filter(caseDetails -> shareCases.localAuthority.equals(caseDetails.getData().get("caseLocalAuthority")))
+            .collect(Collectors.toList());
+
+        log.info("Share case. Found {} case(s) for '{}' LA", casesDetails.size(), shareCases.localAuthority);
+
+        final String serviceToken = authTokenGenerator.generate();
+
+        casesDetails.stream()
+            .forEach(caseDetails ->
+                shareCases.usersIds.forEach(userId -> {
```

Comment:
This part is very similar to the above method. Maybe there could be a private method 
```java
    private void shareCase(List<String> ids, String caseId, String serviceToken) {
        ids.forEach(id -> {
            caseUser.updateCaseRolesForUser(
                requestData.authorisation(),
                serviceToken,
                caseId,
                id,
                new CaseUser(id, CASE_ROLES)
            );
            log.info("Share case. User {} granted access to {}", id, caseId);
        });
    } 
```

Reply:
Then the two controller methods can use that

Reply:
done